src/acl_mem.cpp

```diff
@@ -7527,6 +7790,158 @@ void *acl_get_physical_address(cl_mem mem, cl_device_id device) {
+    return result;
+  }
+
+  acl_device_op_queue_t *doq = &(acl_platform.device_op_queue);
+  acl_device_op_t *last_op = 0;
+
+  // Precautionary, but it also nudges the device scheduler to try
+  // to free up old operation slots.
+  acl_forget_proposed_device_ops(doq);
+
+  last_op = acl_propose_device_op(doq, ACL_DEVICE_OP_DEVICE_GLOBAL_READ, event);
+
+  if (last_op) {
+    // We managed to enqueue everything.
+    event->last_device_op = last_op;
+    acl_commit_proposed_device_ops(doq);
+    result = 1;
+  } else {
+    // Back off, and wait until later when we have more space in the
+    // device op queue.
+    acl_forget_proposed_device_ops(doq);
+  }
+  acl_print_debug_msg(
+      "Exiting acl_submit_read_device_global_device_op function\n");
+  return result;
+}
+
+// Submit a device global write device operation to the device op queue
+cl_int acl_submit_write_device_global_device_op(cl_event event) {
+  acl_print_debug_msg(
+      "Entering acl_submit_write_device_global_device_op function\n");
+  int result = 0;
+  acl_assert_locked();
+
+  // No user-level scheduling blocks this device global write
+  // So submit it to the device op queue.
+  // But only if it isn't already enqueued there.
+  if (!acl_event_is_valid(event)) {
+    return result;
+  }
+  // Already enqueued.
+  if (event->last_device_op) {
+    return result;
+  }
+
+  acl_device_op_queue_t *doq = &(acl_platform.device_op_queue);
+  acl_device_op_t *last_op = 0;
+
+  // Precautionary, but it also nudges the device scheduler to try
+  // to free up old operation slots.
+  acl_forget_proposed_device_ops(doq);
+
+  last_op =
+      acl_propose_device_op(doq, ACL_DEVICE_OP_DEVICE_GLOBAL_WRITE, event);
+
+  if (last_op) {
+    // We managed to enqueue everything.
+    event->last_device_op = last_op;
+    acl_commit_proposed_device_ops(doq);
+    result = 1;
+  } else {
+    // Back off, and wait until later when we have more space in the
+    // device op queue.
+    acl_forget_proposed_device_ops(doq);
+  }
+  acl_print_debug_msg(
+      "Exiting acl_submit_write_device_global_device_op function\n");
+  return result;
+}
+
+// Read from a device global
+void acl_read_device_global(void *user_data, acl_device_op_t *op) {
+  acl_print_debug_msg("Entering acl_read_device_global function\n");
+  cl_event event = op->info.event;
+  cl_int status = 0;
+  size_t pulled_data = 0;
+
+  acl_assert_locked();
+
+  if (!acl_event_is_valid(event) ||
+      !acl_command_queue_is_valid(event->command_queue)) {
+    acl_set_device_op_execution_status(op, -1);
+    return;
+  }
+
+  acl_set_device_op_execution_status(op, CL_SUBMITTED);
+  acl_set_device_op_execution_status(op, CL_RUNNING);
+
+  status = acl_get_hal()->simulation_device_global_interface_read(
+      event->cmd.info.device_global_info.physical_device_id,
+      event->cmd.info.device_global_info.name,
+      event->cmd.info.device_global_info.read_ptr,
+      (size_t)event->cmd.info.device_global_info.device_global_addr,
```

Comment:
Oooo sorry just got another question, should `offset` be added to the `device_global_addr` here?

Reply:
My bad. Fixed.